### PR TITLE
feat: added wire.script.tools

### PR DESCRIPTION
**Pull request by GregoryIvo**
Signed-off-by: Gregory Ivo <gregorybivo@gmail.com>

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

With a single .DP we have added two new replacements for the old Nashorn-based conditional and filter script components.

**Related Issue:** N/A

**Description of the solution adopted:** We Used Graal's Polygot API to make this PR work. 

**Screenshots:** N/A

**Any side note on the changes made:** @marcellorinaldo  co-authored this PR. Run-time maven dependencies:
- `org.graalvm.js/js/22.0.0.2`
- `org.graalvm.js/js-scriptengine/22.0.0.2`
- `org.graalvm.js/js-launcher/22.0.0.2`
- `org.graalvm.sdk/graal-sdk/22.0.0.2`
- `org.graalvm.truffle/truffle-api/22.0.0.2`

### CODE REVIEW

**Comment by GregoryIvo** [comment]
@marcellorinaldo What do you think?

**Comment by marcellorinaldo** [comment]
TODO: update NOTICE file.

**Comment by marcellorinaldo** [comment]
Pending EMO approval for org.graalvm.js/js/22.0.0.2 https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/5271#note_1048938

File: kura/test/org.eclipse.kura.wire.script.tools.test/build.properties
Comment (marcellorinaldo) [inline kura/test/org.eclipse.kura.wire.script.tools.test/build.properties]: Please correct copyright header

File: kura/test/org.eclipse.kura.wire.script.tools.test/pom.xml
Comment (marcellorinaldo) [inline kura/test/org.eclipse.kura.wire.script.tools.test/pom.xml]: Please fix indentation

**Review by marcellorinaldo — CHANGES_REQUESTED** [review]

**Comment by GregoryIvo** [comment]
Updated Example Script:

Script:
```
// the following global variables are available:
// input: an object that represents the received wire envelope
// output: an object that can be used for emitting records
// logger: a slf4j logger

// input has two properties:
// input.emitterPid contains the emitterPid of the received envelope
// input.getRecords() is an immutable array that contains the received records

logger.info('Emitter pid is {}', input.getEmitterPid())

var record = input.getRecords()[0].getProperties() // get the properties of the first record, it is assumed it comes from a Timer


// input records are immutable
// the properties of a record are TypedValues
logger.info('timer value: {}', record.TIMER.getValue()) // print the timer value
logger.info('timer value type: {}', record.TIMER.getType()) // print the timer value type



var records = input.getRecords()

for(let i=0; i<records.length; i++) {
    // WireRecord.getProperties() returns a map of String -> TypedValue
    for (const [keyString, typedValue] of records[i].getProperties()) {
        logger.info('The {}-th record contains:', String(i))
        logger.info('\t{}={}\n', keyString, typedValue.getValue())
    }
}

// creating a list of wire records to emit
var output = new Array()

// each WireRecord contains a map of properties
var outputMap = new Object()

// first record contains an integer
outputMap = new Object()
outputMap['example.integer'] = newIntegerValue(10)
outputMap['example.long'] = newLongValue(100)
outputMap['example.float'] = newFloatValue(1.014)
outputMap['example.double'] = newDoubleValue(10.12)
outputMap['example.boolean'] = newBooleanValue(true)
output[0] = newWireRecord(outputMap)

// second record contains a byte-array
var byteArray = newByteArray(4) 
byteArray[0] = 1
byteArray[1] = 2
byteArray[2] = 0xaa
byteArray[3] = 0xbb
outputMap = new Object()
outputMap['example.byte.array'] = newByteArrayValue(byteArray)
output[1] = newWireRecord(outputMap)

// third record contains strings
outputMap = new Object()
outputMap['example.string1'] = newStringValue('Hello World!')
outputMap['example.string3'] = newStringValue('Hello World Twice!')
outputMap['example.string2'] = newStringValue('Hello World Thrice!')
output[2] = newWireRecord(outputMap)


// the script context is retained across different
// invocations for a single ScriptFilter
// this allows stateful computations

// create a persistent counter
counter = typeof(counter) === 'undefined' ? 0 : counter
counter++

// emit the counter value in a different WireRecord
outputMap = new Object()
outputMap['counter'] = newIntegerValue(counter)
var counterRecord = newWireRecord(outputMap)
output.push(counterRecord)


logger.info('Records that will be emitted: {}', output.length)
```

Output:
```
2022-12-02T17:01:05,696 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger - Record List content:
2022-12-02T17:01:05,696 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -   Record content:
2022-12-02T17:01:05,696 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.integer : 10
2022-12-02T17:01:05,696 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.long : 100
2022-12-02T17:01:05,697 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.float : 1.014
2022-12-02T17:01:05,697 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.double : 10.12
2022-12-02T17:01:05,697 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.boolean : true
2022-12-02T17:01:05,697 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -   Record content:
2022-12-02T17:01:05,697 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.byte.array : [1, 2, 0, 0]
2022-12-02T17:01:05,697 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -   Record content:
2022-12-02T17:01:05,698 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.string1 : Hello World!
2022-12-02T17:01:05,698 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.string3 : Hello World Twice!
2022-12-02T17:01:05,698 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     example.string2 : Hello World Thrice!
2022-12-02T17:01:05,698 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -   Record content:
2022-12-02T17:01:05,698 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -     counter : 231
2022-12-02T17:01:05,698 [WiresTimer_timer_1767873417] INFO  o.e.k.i.w.l.Logger -
```

**Comment by MMaiero** [comment]
Tests are either not picked up or not covering all the new code.

**Comment by GregoryIvo** [comment]
@MMaiero seems to be not picked up. Tried to modify the sonar path in the main Pom, is there any other potential reason why this could happen? 

Edit: I believe I see where I went wrong and corrected it. 🤞

**Comment by marcellorinaldo** [comment]
Documentation: https://github.com/eclipse/kura/pull/4287

**Review by marcellorinaldo — DISMISSED** [review]

**Comment by marcellorinaldo** [comment]
Waiting for EMO approval: https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/5271#note_1048938

**Comment by GregoryIvo** [comment]
@marcellorinaldo made the changes you suggested, please take a look!

File: kura/org.eclipse.kura.wire.script.tools/src/main/java/org/eclipse/kura/wire/script/tools/conditional/component/ConditionalComponent.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.script.tools/src/main/java/org/eclipse/kura/wire/script/tools/conditional/component/ConditionalComponent.java:123]: This condition is never tested, i think it is an important test case.

File: kura/org.eclipse.kura.wire.script.tools/src/main/java/org/eclipse/kura/wire/script/tools/conditional/component/ConditionalComponent.java
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.wire.script.tools/src/main/java/org/eclipse/kura/wire/script/tools/conditional/component/ConditionalComponent.java]: Also this condition is never tested. Can we add some cases on that too?

**Review by marcellorinaldo — CHANGES_REQUESTED** [review]

**Comment by GregoryIvo** [comment]
note: we also need to fix the default script

**Comment by marcellorinaldo** [comment]
> note: we also need to fix the default script

I've added the new default with this commit https://github.com/eclipse/kura/pull/4286/commits/599b92bfa4b9338c25bddb12b2f4db20bc3b59f7. I've tested it working, isn't it anymore?

**Review by marcellorinaldo — APPROVED** [review]

**Comment by marcellorinaldo** [comment]
NOTICE file update: https://github.com/eclipse/kura/pull/4288

**Pull request merged**
